### PR TITLE
[clang-tidy]avoid bugprone-unused-return-value false positive for assignment operator overloading

### LINKED ISSUE #84480: [clang-tidy] bugprone-unused-return-value false positive will matching assignment
It seems like this is a bigger issue.

If you use `::.*` for `CheckedFunctions` to match all functions it will report the warning on *every* assignment (`=` and `+=`) regardless of type.

_Originally posted by @firewave in https://github.com/llvm/llvm-project/issues/84314#issuecomment-1984026830_
            

Here's an example:

```cpp
#include <string>

class C {
public:
    void setS(const char* s) {
        mS = s;
    }
private:
    std::string mS;
};
```

```
test.cpp:6:9: error: the value returned by this function should not be disregarded; neglecting it may lead to errors [bugprone-unused-return-value,-warnings-as-errors]
    6 |         mSt = s;
      |         ^~~~~~~
```

(No godbolt link as I could not figure out how to make the passing of the option work)

_Originally posted by @firewave in https://github.com/llvm/llvm-project/issues/84314#issuecomment-1984118735_

### PR DESCRIPTION
**Pull request by HerrCai0907**
Fixes: #84480
We assuem assignemnt at most of time, operator overloading means the value is assigned to the other variable, then clang-tidy should suppress warning even if this operator overloading match the regex.

### CODE REVIEW

**Comment by firewave** [comment]
Maybe add `+=` to the tests as well? I have also seen it reported with that.

File: clang-tools-extra/docs/clang-tidy/checks/bugprone/unused-return-value.rst [hunk 1]
```diff
@@ -5,6 +5,8 @@ bugprone-unused-return-value
 
 Warns on unused function return values. The checked functions can be configured.
 
+Operator overloading with assignment semantics are ignored。
```
Comment (5chmidti) [inline clang-tools-extra/docs/clang-tidy/checks/bugprone/unused-return-value.rst:8]: The dot at the end of this line is some character that is not the normal dot.

File: clang-tools-extra/clang-tidy/bugprone/UnusedReturnValueCheck.cpp [hunk 2]
```diff
@@ -28,6 +30,11 @@ AST_MATCHER_P(FunctionDecl, isInstantiatedFrom, Matcher<FunctionDecl>,
   return InnerMatcher.matches(InstantiatedFrom ? *InstantiatedFrom : Node,
                               Finder, Builder);
 }
+
+AST_MATCHER_P(CXXMethodDecl, isOperatorOverloading,
+              llvm::SmallVector<OverloadedOperatorKind>, kinds) {
```
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/UnusedReturnValueCheck.cpp:35]: `Kinds`

**Review by 5chmidti — APPROVED** [review]

**Comment by 5chmidti** [comment]
*(Approval after fixes)

**Pull request merged**